### PR TITLE
feat(text-input): add read-only variant

### LINKED ISSUE #1672: [Text input] Add character limit/counter
### Summary

Text fields and text areas may have a maximum number of characters. This character limit might need to be communciated: 
1. to users ahead of time, so they can format their input,
2. Whilst a user is entering their input so they know how many characters remain
3. In the form of a validation error when the user reaches the limit

Design usage guidelines and/or enhancements to the components to support that design would be desirable. 

### Justification

It's a common pattern which you can see on a number of IBM products and tools and would be good to standardise. 

### Desired UX and success metrics

When a user is typing into a form field which has a maximum value, that value is communicated to the user, and if the limit is reached, they are informed whilst typing. 

### "Must have" functionality
Tell the user ahead of time what the value is
Tell the user clearly when they have hit the limit (and stop them entering text) 

Might also include a "x characters remaining" indicator so users know how many more they can enter.

### LINKED ISSUE #3018: Read only text input
ref: https://github.com/carbon-design-system/carbon/issues/2177 by @shixiedesign 

A read only variant of Text input is suggested here: https://github.com/carbon-design-system/carbon-contribution/issues/13#issue-421692303

> The “Disabled” option is not viable, as the label and entered data is not accessible. Proposal is to simply remove the line at the base of the "Filled" version of the component, so that it closely matches with the Disabled component, but passes accessibility.

## Overview of component (updated May 13)
- Read-only variant will be keeping underline and dropping the background of the field. We need the underline for low-vision users to recognize this as a text field. Also loosing the underline makes the element a button-style, which would be confusing. 
- Not moving the text position after all. Keeping text 16px away from left of border helps with preserving the form structure. 
- Regular text cursor and allows text selecting
- Add `not-editable` icon, icon in PR https://github.com/carbon-design-system/carbon/pull/2710
- Optional tooltip when hovering `not-editable` icon, to provide reason for the field being read-only 
- Overflow text in the field will show in tooltip on hover
- Should be considered a state, possible `readonly` attribute
- Follow the standard readonly attribute: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#attr-readonly

## Design spec (updated May 13)

image

## In context explorations: 
image
image

## Finalized design spec. Ready for dev 🤖 updated May 23

Read-only 1

### Light variant spec
Only difference from default is changing field background to `$field-01` so it is the same as background color. 

Read-only 2

### PR DESCRIPTION
**Pull request by emyarod**
Closes #3018

Closes #1672

This PR adds read-only support to the text input and textarea components alongside a mechanism to convert the inputs to controlled mode based on the `onChange` handler and `value` prop. This PR also reimplements character counter (which was previously reverted but depends on the controlled/uncontrolled separation). Since controlled/uncontrolled separation itself doesn’t provide any benefit from user’s perspective until character counter is introduced, it is being introduced alongside the character counter feature here

#### Changelog

**New**

- mechanism to turn `TextInput` and `TextArea` to controlled components based on `onChange` handler and `value` prop
- `readOnly` mode styles
- character counter for `TextInput` and `TextArea` (ref: https://github.com/carbon-design-system/carbon/pull/2745 which was previously reverted)

#### Testing / Reviewing

Ensure the read-only text input and character counters behave as expected

### CODE REVIEW

File: packages/react/src/components/TextInput/TextInput.js
Comment (jendowns) [inline packages/react/src/components/TextInput/TextInput.js]: Just a heads up -- it looks like this propType documentation is mixed up with the `readOnly` propType info on line 267 🕵️‍♀️

Comment (emyarod) [reply packages/react/src/components/TextInput/TextInput.js]: fixed now, thanks!

**Review by shixiedesign — COMMENTED** [review]
Looks great overall! A few small things: 
- [ ] clicking outside the tooltip should close the tooltip (vanilla only, react looks fine) 
Jun-19-2019 10-13-19
- [x] new tooltip should show above existing tooltip on screen, potentially not an issue if we can close the other tooltip easily
Jun-19-2019 10-16-50
image
- [x] Is it possible to disable "error state" on read-only? This situ shouldn't arise. It doesn't give user a way forward since field cannot be edited. 
image

**Comment by emyarod** [comment]
> clicking outside the tooltip should close the tooltip (vanilla only, react looks fine)

@shixiedesign clicking outside of the tooltip in both vanilla and react closes the tooltip for me. can you provide more details about how to reproduce this issue?

**Comment by elizabethsjudd** [comment]
@emyarod Depending on what version of the tooltip you are using it could be related to this question i had: https://github.com/carbon-design-system/carbon/issues/3099#issuecomment-503701949

**Comment by emyarod** [comment]
@elizabethsjudd this is using the WCAG 2.0 compliant tooltip. do you mean the behavior is inconsistent when a button is the trigger element?

**Comment by elizabethsjudd** [comment]
@emyarod when a button is used as the trigger for a tooltip, the user can "click" on the button and the tooltip will remain active until the user clicks somewhere else (the button is given focus which also causes the tooltip to display).  You can scroll, resize, mouse over other elements, etc all while  the tooltip remains open (this is what I believe is causing the scenario @shixiedesign is pointing out by having two tooltips open at once).  The one thing I don't know is if this is accurate behavior because in actual offerings we'd expect a button to perform an action which would most likely cause the focus to change/move and therefore I believe the tooltip would hide.  The question that I'm bringing up however is that I feel like both of these tooltips seem like they should be "definition tooltips" as they are either explaining the reason something is in a state, or it's just the extended version of the inputs value.  My question in the other issue is that I don't feel like "definition tooltips" should be clickable at all unless they are on a button that is actually performing another action.  In this use-case I don't feel that the tooltips should be on clickable triggers and the clicking is what is making them persistent. I've done a ton of research and implementations of various types of a11y tooltips so if you have any questions around this don't hesitate to ask.

**Comment by emyarod** [comment]
@elizabethsjudd that was intentional according to the initial component spec so that the user can select the content (although clicking on the read only input will also select all of the text). from what I understand, @shixiedesign is referring to clicking multiple times outside of the button before the tooltip closes, but I am unable to replicate that behavior (the tooltip is dismissed with one outside click for me). I think I need some clarification on what the bug report is referring to and some confirmation on the expected behavior

**Comment by elizabethsjudd** [comment]
@emyarod thanks for the clarification.  Just out of curiosity, are you using the first tooltip that's on this page: https://www.carbondesignsystem.com/components/tooltip/code?  At least on the site demo even after I open the tooltip if I try to select the text the tooltip closes (because it closes by clicking anywhere outside the `trigger`).  WH found this when we were working with tooltips and thought this was odd behavior. Is that an update that's also coming with this PR?

**Comment by emyarod** [comment]
@elizabethsjudd, no I don't make any changes to the tooltip code in this PR so it seems you found a vanilla-only interactive tooltip bug. it looks like if I click on one of the interactive elements in the tooltip before trying to select the text the tooltip will not close (but only in the dev environment, not the Carbon website). probably need to examine the vanilla tooltip click handler. we can track this in a separate issue

**Comment by elizabethsjudd** [comment]
@emyarod totally agree that it is out of scope for this issue.  Thank you for confirming!

**Comment by elizabethsjudd** [comment]
@emyarod I was just thinking about it. using `aria-haspopup="true"` would actually probably fix the "confusion" aspect that I was talking about earlier.  Then the `aria-controls` applied dynamically should still allow JAWs users to toggle between the two when the the tooltip is open.

**Comment by emyarod** [comment]
@elizabethsjudd just to confirm, `aria-haspopup="true"` would be persistent and `aria-controls` would be added depending on if the tooltip content is inserted in the DOM right?

also is this related to https://github.com/carbon-design-system/carbon/issues/3100?

**Comment by elizabethsjudd** [comment]
@emyarod yes, and I would say that #3100 is related.  I have a PR open for #3100 right now (see #3148). FYI though, #3148 only resolves AVT 1 issues. There are still issues with AVT 2 & 3 that I'm hoping to address next.

**Comment by emyarod** [comment]
@shixiedesign I think this is good for another review before I port this over to React but I may hold off until #3148 and associated PRs are approved and merged

**Review by shixiedesign — APPROVED** [review]
Looks good!

File: packages/react/src/components/TextArea/TextArea.js
Comment (asudoh) [inline packages/react/src/components/TextArea/TextArea.js]: Probably `useCharCount`? (To distinguish it from the count (length) itself)

File: packages/react/src/components/TextInput/TextInput.js
Comment (asudoh) [inline packages/react/src/components/TextInput/TextInput.js]: Probably `useCharCount`? (To distinguish it from the count (length) itself)

File: packages/react/src/components/TextArea/TextArea.js
Comment (asudoh) [inline packages/react/src/components/TextArea/TextArea.js:303]: Need `onChange` prop, too, with "missing `onChange` in controlled mode" check.

File: packages/react/src/components/TextInput/TextInput.js
Comment (asudoh) [inline packages/react/src/components/TextInput/TextInput.js:291]: Need prop type of `onChange`, too, with "missing `onChange` in controlled mode" check.

File: packages/react/src/components/TextInput/TextInput.js
Comment (asudoh) [inline packages/react/src/components/TextInput/TextInput.js:96]: We need to add `readOnly` to `<TextArea>`, too.

**Review by asudoh — COMMENTED** [review]
Nice progress @emyarod!

File: packages/react/src/components/TextInput/TextInput.js
Comment (emyarod) [reply packages/react/src/components/TextInput/TextInput.js:96]: I was thinking of separating the character counter work from the read-only work unless it's fine to contain it all in this PR. if so, I will add read-only to `<TextArea>` before marking this draft as ready

Comment (asudoh) [reply packages/react/src/components/TextInput/TextInput.js:96]: If adding read-only to `<TextArea>` would cause much code it'll make sense to separate it. Otherwise including the change in this would be fine.

File: packages/react/src/components/TextArea/TextArea.js
Comment (emyarod) [reply packages/react/src/components/TextArea/TextArea.js]: sure, renamed this for `TextInput` and `TextArea`

File: packages/react/src/components/TextInput/TextInput.js
Comment (emyarod) [reply packages/react/src/components/TextInput/TextInput.js:96]: got it, just added it to vanilla and React

Comment (emyarod) [reply packages/react/src/components/TextInput/TextInput.js:96]: we will need additional design support for readonly textarea though, I don't believe we have a spec for that

**Comment by emyarod** [comment]
I believe this will be ready for review once the questions around readonly `TextArea` spec are addressed

**Comment by joshblack** [comment]
@emyarod anything we can do to help move this along or just waiting for what you pointed out?

**Comment by emyarod** [comment]
@joshblack only waiting for spec decisions otherwise this has been ready for dev review

**Comment by aagonzales** [comment]
Why is it jumping when I toggle between the states? Is this just an issue in the story book?
Jul-09-2019 11-25-50


Also does it for `invalid`, I'm guessing its because of the icon? Will it alway jump? If yes, then ideally only the right side would be changing, the type shouldn't be moving.

**Comment by emyarod** [comment]
> Is this just an issue in the story book?

@aagonzales yeah pretty much. the reason the width changes for invalid and readonly states is because we add padding to the textarea so that the text won't overlap with the status icons. we don't set a `width` rule for the component out of the box so it will increase in size as needed/as its container will allow. the vanilla examples kind of illustrate this, all of the examples have the same width

File: packages/react/src/components/TextArea/TextArea.js
Comment (emyarod) [reply packages/react/src/components/TextArea/TextArea.js:303]: this should be addressed now

File: packages/react/src/components/TextInput/TextInput.js
Comment (emyarod) [reply packages/react/src/components/TextInput/TextInput.js:291]: this should be addressed now

**Review by designertyler — APPROVED** [review]

File: packages/react/src/components/TextArea/TextArea.js
Comment (joshblack) [inline packages/react/src/components/TextArea/TextArea.js:84]: What's the reasoning behind the `null` case? Is it from passing through to `DefaultCharCounter` and us not wanting to render `undefined`?

File: packages/react/src/components/TextArea/TextArea.js
Comment (joshblack) [inline packages/react/src/components/TextArea/TextArea.js]: Would we just want to pull this out into its own component like `DefaultCharCounter`?

File: packages/react/src/components/TextArea/TextArea.js
Comment (joshblack) [inline packages/react/src/components/TextArea/TextArea.js]: Same question as for label

File: packages/react/src/components/TextArea/TextArea.js
Comment (joshblack) [inline packages/react/src/components/TextArea/TextArea.js]: Do we need to cast to `null`? Curious what the behavior with `undefined` is in this case

File: packages/react/src/components/TextArea/TextArea.js
Comment (joshblack) [inline packages/react/src/components/TextArea/TextArea.js:153]: Do we need to worry about `other.defaultValue` being spread to this?

File: packages/components/src/components/text-area/_text-area.scss
Comment (asudoh) [inline packages/components/src/components/text-area/_text-area.scss]: Can we add a reset style to negate focus outline for read-only variant?

File: packages/react/src/components/TextInput/TextInput.js
Comment (joshblack) [inline packages/react/src/components/TextInput/TextInput.js]: Is there any way we can share some of this similar (at a glance) logic between TextArea and TextInput?

**Review by joshblack — CHANGES_REQUESTED** [review]

**Review by asudoh — COMMENTED** [review]
Just one thing - Thanks @emyarod!

File: packages/react/src/components/TextArea/TextArea.js
Comment (emyarod) [reply packages/react/src/components/TextArea/TextArea.js]: this was just out of habit, since I think for `readOnly`,  `null` and `undefined` behave the same way, but for the aria attributes above, they differ

File: packages/react/src/components/TextInput/TextInput.js
Comment (emyarod) [reply packages/react/src/components/TextInput/TextInput.js]: it may be worth exploring once password input is finalized since it shares even more with TextInput than TextArea

File: packages/react/src/components/TextArea/TextArea.js
Comment (emyarod) [reply packages/react/src/components/TextArea/TextArea.js:153]: no I don't think so, since it's destructured

File: packages/react/src/components/TextArea/TextArea.js
Comment (emyarod) [reply packages/react/src/components/TextArea/TextArea.js]: I kept this for the sake of consistency with all of our other form components

File: packages/react/src/components/TextArea/TextArea.js
Comment (emyarod) [reply packages/react/src/components/TextArea/TextArea.js]: I kept this for the sake of consistency with all of our other form components

File: packages/react/src/components/TextArea/TextArea.js
Comment (emyarod) [reply packages/react/src/components/TextArea/TextArea.js:84]: yeah it also deals with the case where maxLength is 0

File: packages/components/src/components/text-area/_text-area.scss
Comment (emyarod) [reply packages/components/src/components/text-area/_text-area.scss]: sure, added reset styles for read-only

**Review by asudoh — APPROVED** [review]
🎉 Thank you for the update @emyarod!

File: packages/react/src/components/TextArea/TextArea.js
Comment (joshblack) [reply packages/react/src/components/TextArea/TextArea.js]: We really shouldn't be doing an IIFE block though, right? Feels off to create a whole closure for the sake of what seems to be a conditional

Comment (emyarod) [reply packages/react/src/components/TextArea/TextArea.js]: I don't see a reason not to use IIFE here since the closure keeps `labelContent` private and only `label` needs to access it

File: packages/react/src/components/TextArea/TextArea.js
Comment (asudoh) [inline packages/react/src/components/TextArea/TextArea.js]: Oops forgot to mention that, my apologies - For `<TextInput>` and `<TextArea>` we don't have a breaking change, and thus the feature flag won't be needed.

Comment (emyarod) [reply packages/react/src/components/TextArea/TextArea.js]: oh yeah we discussed this previously, reverted the commit

Comment (asudoh) [reply packages/react/src/components/TextArea/TextArea.js]: @emyarod I meant `onChange: requiredIfValueExists(PropTypes.func)` - Wondering if that's what you did.

Comment (emyarod) [reply packages/react/src/components/TextArea/TextArea.js]: yup updated @asudoh

Comment (asudoh) [reply packages/react/src/components/TextArea/TextArea.js]: Cool thanks @emyarod!

**Comment by asudoh** [comment]
@joshblack Please let us know if any of your questions haven't been addressed yet. Thanks!

**Comment by joshblack** [comment]
@asudoh have been in the process of running the endgame and project planning, can circle back later this week. Also might help to either break up this PR into smaller parts, or schedule a meeting for the team to go over it 👍

**Comment by asudoh** [comment]
@emyarod Would you want to schedule one? Thanks!

**Comment by joshblack** [comment]
We should defer on this till the PR review meeting so that the team can take a look 👍 Also splitting this up into smaller chunks can help with the review process.

**Comment by asudoh** [comment]
This PR seems with fairly containable size. While review meeting may be a good idea, I'd urge for due diligence of reviewing the code before it happens, probably as a homework - As long as you still are interested in reviewing.

**Review by abbeyhrt — APPROVED** [review]

**Review by joshblack — APPROVED** [review]

File: packages/react/src/components/TextInput/TextInput.js
Comment (asudoh) [inline packages/react/src/components/TextInput/TextInput.js]: Is the usage of interactive tooltip here from the UI spec our designer gave us?

**Review by asudoh — COMMENTED** [review]
I already signed off, but just one thing before merging - Thank you for all the updates 🎉 @emyarod!

File: packages/react/src/components/TextInput/TextInput.js
Comment (emyarod) [reply packages/react/src/components/TextInput/TextInput.js]: yeah the tooltip is for showing the full value of the input field when it is in read only mode (ref https://github.com/carbon-design-system/carbon/issues/2177)

Comment (asudoh) [reply packages/react/src/components/TextInput/TextInput.js]: Thanks, looking at https://user-images.githubusercontent.com/15144993/57665202-9970fc00-75c0-11e9-9a3f-b628d7b168a9.png - Probably my specific question is to double-check if it's an interactive tooltip, instead of the hover-over variant (definition/icon)?

Comment (emyarod) [reply packages/react/src/components/TextInput/TextInput.js]: the reason for interactive tooltip is because the content could be several lines long, which isn't suggested for icon or definition tooltips IIRC. but we may need to make some updates due to #3812

**Comment by elizabethsjudd** [comment]
@emyarod From an a11y perspective I would expect the read-only text input to use a definition tooltip because it would always be static text.  From my work on the vanilla version of definition tooltip we did make sure it supported multiple lines of text.

**Comment by emyarod** [comment]
@elizabethsjudd sure in that case I need to address https://github.com/carbon-design-system/carbon/issues/3813 first, but I think it was just a semantics thing since I believe there were some guidelines about suggested content length for definitions

**Comment by asudoh** [comment]
CC @shixiedesign just to double-check if interactive tooltip (body) usage is good here from design perspective. Given it seems that interactive tooltip code usage here is just for showing tooltip upon click/focus, from technical perspective it's too much to bring in for read-only text input purpose (e.g. Interactive tooltip is a floating menu whereas read-only text input doesn't require that).

**Comment by joshblack** [comment]
bump @shixiedesign if you have a sec!

**Comment by emyarod** [comment]
I swapped the interactive tooltip with definition tooltip ahead of final design confirmation. but if we decide that showing the tooltip on input hover is not desired, then we can revert the latest commit. other than that, due to recent tooltip changes we should merge https://github.com/carbon-design-system/carbon/pull/3954 ahead of this

**Comment by shixiedesign** [comment]
I will take a look at this again soon. Sorry for the hold up. There's only 24 hours in the day 😞

**Review by asudoh — COMMENTED** [review]
Latest code LGTM 👍 from technical perspective - Thanks @emyarod!

**Comment by shixiedesign** [comment]
What an epic long thread in this PR! I'd love to merge it as is, but found a few possible errors. I used Chrome incognito window: 

- [ ] Icon for "read only" is somehow not appearing on the right side of the input field
- [ ] On react version, I see a blue dotted underline, which is not in spec. If definition tooltip is better for accessibility purposes, can we hide the blue dotted line? I do not see it in vanilla.
image

**Comment by emyarod** [comment]
some of those issues may be related to stale netlify build, but I'll double check. I recall specifically resolving both of those issues while moving from interactive to definition tooltip so I'll take another look

**Comment by emyarod** [comment]
@shixiedesign the screenshot looks outdated, can you recheck? the react version should not have had either of the issues, and the vanilla icon is now correctly placed. neither version should have shown the dotted blue line though so I think it's due to the preview not being up to date

**Review by shixiedesign — COMMENTED** [review]
Just one comment, we need to bring back the focus border! 
image

**Comment by emyarod** [comment]
depends on #3995 #4016

**Comment by jnm2377** [comment]
@emyarod can you explain how this is being blocked by https://github.com/carbon-design-system/carbon/pull/3986 ? It seems like that PR needs a bigger discussion, so I'm wondering how we can get this PR moving along. This is blocking another stale PR.

**Comment by emyarod** [comment]
@jnm2377 I am currently exploring ways to work around it

**Comment by joshblack** [comment]
If we don't have a clear resolution in place, would be best to close this for now and revisit in our next components project 👍

**Comment by joshblack** [comment]
Closing this since the stale warning label was applied >7 days ago 👍

**Review by vpicone — COMMENTED** [review]
I think we could use the [`:read-only` pseudo-selector](https://developer.mozilla.org/en-US/docs/Web/CSS/:read-only) here rather than the attribute selector. Would likely only matter for folks overriding using the most "appropriate" selector.

Do we want to remove all the vanilla stuff from the PR?

File: packages/react/src/components/TextInput/TextInput-story.js
Comment (vpicone) [inline packages/react/src/components/TextInput/TextInput-story.js:133]: Nice work ignoring validation here, we definitely want to make sure that's the case in any readonly inputs

**Pull request closed without merging**
